apps/web/vibes/soul/primitives/dynamic-form/index.tsx

```diff
@@ -0,0 +1,275 @@
+}
+
+export type DynamicFormAction<F extends Field> = Action<State<F>, FormData>;
+
+interface Props<F extends Field> {
+  fields: F[];
+  action: DynamicFormAction<F>;
+  syncQueryParams?: boolean;
+  ctaLabel?: string;
+}
+
+export function DynamicForm<F extends Field>({
+  action,
+  fields,
+  ctaLabel = 'Submit',
+  syncQueryParams = false,
+}: Props<F>) {
+  const [params] = useQueryStates(
+    fields.reduce<Record<string, typeof parseAsString>>(
+      (acc, field) => {
+        return { ...acc, [field.name]: parseAsString };
+      },
+      { quantity: parseAsString },
+    ),
+    { shallow: false },
+  );
+  const defaultValue = fields.reduce<{
+    [Key in keyof SchemaRawShape]?: z.infer<SchemaRawShape[Key]>;
+  }>(
+    (acc, field) => ({
+      ...acc,
+      [field.name]: (syncQueryParams ? params[field.name] : null) ?? field.defaultValue ?? '',
+    }),
+    { quantity: 1 },
+  );
+  const [{ lastResult }, formAction] = useActionState(action, {
+    fields,
+    lastResult: null,
+  });
+  const [form, formFields] = useForm({
+    lastResult,
+    constraint: getZodConstraint(schema(fields)),
+    onValidate({ formData }) {
+      return parseWithZod(formData, { schema: schema(fields) });
+    },
+    // @ts-expect-error: `defaultValue` types are conflicting with `onValidate`.
+    defaultValue,
+    shouldValidate: 'onSubmit',
+    shouldRevalidate: 'onInput',
+  });
+
+  return (
+    <FormProvider context={form.context}>
+      <form {...getFormProps(form)} action={formAction}>
+        <div className="space-y-6">
+          {fields.map((field) => {
+            return (
+              <DynamicFormField
+                field={field}
+                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                formField={formFields[field.name]!}
+                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                key={formFields[field.name]!.id}
+                syncQueryParams={syncQueryParams}
+              />
+            );
+          })}
+          <div className="flex gap-x-3 pt-3">
+            <SubmitButton>{ctaLabel}</SubmitButton>
+          </div>
+        </div>
+      </form>
+    </FormProvider>
+  );
+}
+
+function SubmitButton({ children }: { children: React.ReactNode }) {
+  const { pending } = useFormStatus();
+
+  return (
+    <Button className="w-auto @xl:w-56" loading={pending} size="medium" type="submit">
+      {children}
+    </Button>
+  );
+}
+
+function DynamicFormField({
+  field,
+  formField,
+  syncQueryParams,
+}: {
+  field: Field;
+  formField: FieldMetadata<string | number | boolean | Date | undefined>;
+  syncQueryParams?: boolean;
+}) {
+  const controls = useInputControl(formField);
+  const [, setParam] = useQueryState(field.name, parseAsString.withOptions({ shallow: false }));
+
+  const handleChange = useCallback(
+    (value: string) => {
+      if (syncQueryParams === true) void setParam(value);
```

Comment:
I think we should use transitions to handle these async calls together with handling pending states